### ISSUE TITLE
Make enqueue conditional based on need/location

### ISSUE THREAD

**Issue by jtsternberg** [WebDevStudios/WP-Contributions#1]
This css should only be enqueued if it's needed: https://github.com/WebDevStudios/WP-Contributions/blob/master/wp-contributions.php#L115-L120

**Comment by jtsternberg** [comment]
Hmm challenge here is knowing if a widget is going to be visible on the page and still getting the css to load in the head (so that you don't have FOUT).

**Comment by dustyf** [comment]
Agreed. Was already thinking of how to do it. I also am document in the template tags for people to use so I would also probably want it if those are used. Any suggestions welcome.

**Comment by colorful-tones** [comment]
I'm all about optimized performance (i.e. minimizing unnecessary requests), but almost seems negligible in this instance. Especially since it is a local (native WP) resource.

With that said, I'd be very interested in method for checking to enqueue.

**Comment by jtsternberg** [comment]
It's 2 more http requests which is the main bummer. Dashicons being native to WP doesn't mean they are loaded for free. BUT I agree that the savings are minimal considering the magic that will likely have to happen to check for the widgets while still in <head>. But I hear @dustyf's good with magic.

**Comment by colorful-tones** [comment]
Wait, I'm confused. We're loading stylesheet regardless of Dashicons, because we need to style stuff either way. So wouldn't this be one request we're talking about?

**Comment by jtsternberg** [comment]
Per the original comment, the plugin's stylesheet (and its dashicons dependency), should only be enqueued if it's needed. What if the user doesn't add the widget for a while? having the plugin active, they're going to be loading those 2 files anyway. Or what if they only want the widget on their single-post sidebar? They're still getting those files on every page-load across the site.

**Comment by colorful-tones** [comment]
Ah yes, I'm following you now. Thanks for clarifying @jtsternberg

**Comment by jtsternberg** [comment]
:thumbsup:

**Comment by colorful-tones** [comment]
So I'm thinking out loud... maybe we `wp_register_style` instead of `enqueue`, which I should of done in first place so it can be deregistered ([wp_deregister_style](https://codex.wordpress.org/Function_Reference/wp_deregister_style)) if anyone chooses so. And then put in function for boolean value to check in each widget whether it is needed for that widget, and whether or not it is already enqueued? Does that sound practical?

One thing I'm struggling with sorting out is whether I should enqueue the plugin's stylesheet, which could include any Dashicons-related CSS, or whether to have a separate stylesheet for any Dashicons CSS and only include when needed? Probably not worth the extra request.

Maybe we set a variable that can be passed as the [dependency option to wp_enqueue_style](https://codex.wordpress.org/Function_Reference/wp_enqueue_style), and then we can set it/check it in each widget that needs it, as well as have a function that users can drop in their functions.php of theme which will load if they need it in there template? E.g. `add_theme_support( 'wp-contributions' );`

`$dependency = 'dashicons';` because not every widget or template tag will need Dashicons.

Am I making any sense? :)

**Comment by tw2113** [comment]
Upon further reflection, I'm actually going to close this one as wontfix. We have very minimal CSS, and browsers have been better with loading things in parallel and actually prefer that with HTTP2+
